oak/server/rust/oak_runtime/src/runtime.rs

```diff
@@ -0,0 +1,136 @@
+use core::sync::atomic::Ordering::SeqCst;
+
+use log::error;
+
+use oak_abi::OakStatus;
+
+use crate::channel;
+use crate::channel::{ChannelReader, ChannelWriter};
+use crate::node::{load_wasm, NodeConfiguration};
+use crate::platform::*;
+use crate::proto;
+use crate::proto::manager::NodeConfiguration_oneof_config_type;
+
+/// Runtime structure for configuring and running a set of Oak nodes.
+pub struct Runtime {
+    configurations: HashMap<String, NodeConfiguration>,
+    terminating: AtomicBool,
+    node_threads: Mutex<Vec<std::thread::JoinHandle<()>>>,
+}
+
+impl Runtime {
+    /// Configure and run the protobuf specified `ApplicationConfiguration`. After creating a
+    /// `Runtime` calling `stop` will send termination signals to nodes and wait for them to
+    /// terminate.
+    pub fn configure_and_run(
+        app_config: proto::manager::ApplicationConfiguration,
+    ) -> Result<(RuntimeRef, ChannelWriter), OakStatus> {
+        let mut runtime = Runtime {
+            configurations: HashMap::new(),
+            terminating: AtomicBool::new(false),
+            node_threads: Mutex::new(Vec::new()),
+        };
+
+        for config in app_config.get_node_configs() {
+            runtime.configurations.insert(
+                config.name.clone(),
+                match &config.config_type {
+                    None => {
+                        error!("Node config {} with no type", config.name);
+                        return Err(OakStatus::ERR_INVALID_ARGS);
+                    }
+                    Some(NodeConfiguration_oneof_config_type::log_config(_)) => {
+                        NodeConfiguration::LogNode
+                    }
+                    Some(NodeConfiguration_oneof_config_type::wasm_config(
+                        proto::manager::WebAssemblyConfiguration { module_bytes, .. },
+                    )) => load_wasm(&module_bytes),
+                    Some(_) => {
+                        error!("Pseudo-node not implemented!");
+                        return Err(OakStatus::ERR_INVALID_ARGS);
+                    }
+                },
+            );
+        }
+
+        let runtime = RuntimeRef(Arc::new(runtime));
+        let (chan_writer, chan_reader) = channel::new();
+
+        runtime.node_create(
+            &app_config.initial_node_config_name,
+            &app_config.initial_entrypoint_name,
+            chan_reader,
+        )?;
+
+        Ok((runtime, chan_writer))
+    }
+
+    /// Thread safe method for determining if the `Runtime` is terminating.
+    pub fn is_terminating(&self) -> bool {
+        self.terminating.load(SeqCst)
+    }
+
+    /// Thread safe method for signaling termination to a `Runtime` and waiting for it's node
+    /// threads to terminate.
+    pub fn stop(&self) {
+        let mut node_threads = self.node_threads.lock().unwrap();
+
+        self.terminating.store(true, SeqCst);
+
+        let mut handles = vec![];
+        std::mem::swap(&mut handles, &mut node_threads);
```

Comment:
I have never seen `std::mem::swap` used in practice :smiley: 

I'm not sure what the advantage is though; AFAICT the lock is still held until the end of the block, so you may as well just iterate on `node_threads` directly, and then call `clear` on the `Vec`?

Reply:
The issue is `JoinHandle` is non copy/clone, the method `JoinHandle::join` takes `self`, and `Mutex::lock()` returns a reference.

Meaning we can't do the following:
- `node_threads.iter()` - gives us items of `&self` and no way to get to `self`
- `node_threads.into_iter()` - can't move vec out of shared reference (borrow from mutex lock)
- `node_threads.clone()` - can't clone vec because `JoinHandle` can't clone

The `std::mem::swap` trick switches our mutably borrowed vec with a vec that is not borrowed at function scope allowing us to consume the temporary vec with the for loop as it is no longer borrowed (after the for loop `handles` is moved and no longer accessible).

There is a simpler alternative but we can't use it, which is to consume the `Runtime` `self` allowing us to use the (also `self` consuming) `Mutex::into_inner`:

~~~rust
pub fn stop(self) {
    let node_threads = self.node_threads.into_inner().unwrap();

    self.terminating.store(true, SeqCst);

    for handle in node_threads {
        handle.join().expect("Failed to join handle");
    }
}
~~~

Although `oak_runtime` will compile like this, no downstream code can call this method as we `Runtime` is always constructed as an `Arc` to be able to pass references to the runtime to node threads.

I guess its possible to give node threads only `Weak` references and then the runtime creator can call  `try_unwrap` on it's runtime to destroy the `Arc` and then it could call this alternative `Runtime::stop`, but at that point I'm not sure there is an advantage.

Reply:
Got it. Would https://doc.rust-lang.org/beta/std/cell/struct.RefCell.html help here by any chance?

Reply:
It could be used, I believe the swap/replace methods would have the same effect. The other methods of `RefCell` are unnecessary though, e.g. we get the (mut)borrow from the mutex.